js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx

```diff
@@ -1,118 +1,181 @@
+          currentState!.onFetched.push(resolve as any);
         });
+      }
+
+      const state = {onFetched: [] as ((value: any) => void)[]};
+      fetchState[key] = state;
+
+      const queryResult = await client.query<TQuery, TVariables>({
+        query,
+        variables,
+        fetchPolicy: 'no-cache',
       });
-    }
-    fetchState[key] = {onFetched: []};
-    // Use client.query here so that we initially use the apollo cache if any data is available in it
-    // and so that we don't subscribe to any updates to that cache (useLazyQuery and useQuery would both subscribe to updates to the
-    // cache which can be very slow)
-    const queryResult = await client.query<TQuery, TVariables>({
-      query,
-      variables,
-      fetchPolicy: 'no-cache', // Don't store the result in the cache,
-      // should help avoid page stuttering due to granular updates to the data
-    });
-    const {data} = queryResult;
-    setLoading(false);
-    lru.set(
-      'cache',
-      {data, version},
-      {
-        expiry: new Date('3000-01-01'), // never expire,
-      },
-    );
-    delete fetchState[key];
-    const onFetched = fetchState[key]?.onFetched;
-    try {
-      setData(data);
-    } catch (e) {
-      setTimeout(() => {
-        throw e;
-      });
-    }
-    onFetched?.forEach((cb) => {
-      try {
-        cb(queryResult);
-      } catch (e) {
-        setTimeout(() => {
-          throw e;
-        });
+
+      const {data} = queryResult;
+      await cacheManager.set(data, version);
+
+      const onFetchedHandlers = state.onFetched;
+      if (fetchState[key] === state) {
+        delete fetchState[key]; // Clean up fetch state after handling
       }
-    });
 
-    return queryResult;
-  }, [client, key, lru, query, variables, version]);
+      onFetchedHandlers.forEach((handler) => handler(data)); // Notify all waiting fetches
+
+      return data;
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
+  );
+}
+
+export function useGetCachedData() {
+  const {getCacheManager} = useContext(IndexedDBCacheContext);
 
+  return useCallback(
+    async <TQuery,>({key, version}: {key: string; version: number}) => {
+      const cacheManager = getCacheManager<TQuery>(key);
+      return await cacheManager.get(version);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
+  );
+}
+
+const contextValue = createIndexedDBCacheContextValue();
+export const IndexedDBCacheContext = createContext(contextValue);
+
+export function createIndexedDBCacheContextValue() {
   return {
-    fetch,
-    data,
-    loading,
+    getCacheManager: memoize(<TQuery,>(key: string) => {
+      return new CacheManager<TQuery>(key);
+    }),
+    fetchState: {} as Record<
+      string,
+      {
+        onFetched: ((value: any) => void)[];
+      }
+    >,
   };
 }
 
 export const __resetForJest = () => {
-  fetchState = {};
+  Object.assign(contextValue, createIndexedDBCacheContextValue());
```

Comment:
Everything in here seems great but at a really high level, going `IndexedDBCacheContext => getCacheManager => new CacheManager =>  idb-lru-cache library => get` is a lot of layers of abstraction to essentially read/write a single key that next expires and is never evicted.

There are a lot of "manager things" in here and not a lot of actual "management" of the cache happening. 

I think a lot of the complexity arises from the fact that we're using an LRU cache to get/set each key and don't want the "LRU" part, so we have to instantiate a separate copy for each callsite and then deal with the object oriented patterns and instances.  

I think instantiating a single global LRU cache and giving it a "high-enough-to-never-hit" max key count, and then using that instead of `getCacheManager, etc`  might make this simpler. (Or perhaps a different low level library that just implements get/set)